### DISCUSSION TITLE
How to estimate the computational cost of gradient optimization

### DISCUSSION THREAD

**Discussion by plantom007** [mdolab/dafoam#545] [General]
Hi，
I want to estimate the computational cost of gradient optimization. I used the SLSQP algorithm to optimize the airfoil and iterated 30 times. The results showed that the function was evaluated 35 times and the gradient was evaluated 31 times. Are the extra 5 CFD calculations a result of the line search?
We know the computational cost of solving the adjoint equation is comparable with that of flow simulation. Does this mean that the computational cost of this optimization approximates 66 CFD calculations? Are there any other missing calculated costs?
Looking forward to your answer, thank you.
![1703685553900](https://github.com/mdolab/dafoam/assets/55193251/62a272d2-d8d0-4a2e-b90a-a5d8e6a639f0)

**Comment by friedenhe** [comment]
The primal and adjoint runtime is printed to the screen (right after each primal and adjoint solution) when you run DAFoam.

**Reply by plantom007** [reply]
Sorry, maybe I didn't describe it clearly. I am trying to figure out how many times the CFD solver is run and how many times the adjoint solver is run during gradient optimization.
I checked the log file. It performed 40 flow field calculations and 31 sensitivity analyses. Can I consider it equivalent to solving the flow governing equation 40 times and the adjoint equation 31 times? Because I only want to focus on the flow field and the number of calculations of the adjoint solver rather than the runtime.

What I'm confused about is whether the flow solution is recalculated when calculating the sensitivity when using the SLSQP algorithm.
Because gradient optimization based on the SLSQP algorithm requires new flow solutions and adjoint solutions, otherwise the flow field solution and adjoint solution cannot be matched one-to-one. But, with the simple gradient descent method, the number of function calculations and the number of gradient calculations are the same. Therefore, adjoint calculations can be performed directly based on previously calculated flow field solutions.
Maybe my understanding is wrong, I hope you can explain it in detail. Thank you.

**Comment by friedenhe** [comment]
> I checked the log file. It performed 40 flow field calculations and 31 sensitivity analyses. Can I consider it equivalent to solving the flow governing equation 40 times and the adjoint equation 31 times?

Yes.

> What I'm confused about is whether the flow solution is recalculated when calculating the sensitivity when using the SLSQP algorithm. Because gradient optimization based on the SLSQP algorithm requires new flow solutions and adjoint solutions, otherwise the flow field solution and adjoint solution cannot be matched one-to-one. But, with the simple gradient descent method, the number of function calculations and the number of gradient calculations are the same. Therefore, adjoint calculations can be performed directly based on previously calculated flow field solutions. Maybe my understanding is wrong, I hope you can explain it in detail. Thank you.

The reason they are different is that, SLSQP does line searches at each optimization iteration, so it may perform multiple flow solutions until it finds a good step size.
